mlflow/prophet.py

```diff
@@ -0,0 +1,296 @@
+import os
+import yaml
+
+import mlflow
+from mlflow import pyfunc
+from mlflow.utils.requirements_utils import _get_pinned_requirement
+from mlflow.utils.environment import (
+    _mlflow_conda_env,
+    _validate_env_arguments,
+    _process_pip_requirements,
+    _process_conda_env,
+    _CONDA_ENV_FILE_NAME,
+    _REQUIREMENTS_FILE_NAME,
+    _CONSTRAINTS_FILE_NAME,
+)
+from mlflow.utils.file_utils import write_to
+from mlflow.utils.docstring_utils import format_docstring, LOG_MODEL_PARAM_DOCS
+from mlflow.models.signature import ModelSignature
+from mlflow.models.utils import _save_example
+from mlflow.models import Model, ModelInputExample
+from mlflow.tracking.artifact_utils import _download_artifact_from_uri
+from mlflow.utils.model_utils import _get_flavor_configuration
+from mlflow.models.model import MLMODEL_FILE_NAME
+from mlflow.exceptions import MlflowException
+from mlflow.tracking._model_registry import DEFAULT_AWAIT_MAX_SLEEP_SECONDS
+
+FLAVOR_NAME = "prophet"
+_MODEL_BINARY_KEY = "data"
+_MODEL_BINARY_FILE_NAME = "model.pr"
+_MODEL_TYPE_KEY = "model_type"
+_SAVE_FORMAT_KEY = "save_format"
+
+
+def get_default_pip_requirements():
+    """
+    :return: A list of default pip requirements for MLflow Models produced by this flavor.
+             Calls to :func:`save_model()` and :func:`log_model()` produce a pip environment
+             that, at a minimum, contains these requirements.
+    """
+
+    prophet_requirements = [
+        "Cython",
+        "pandas",
+        "convertdate",
+        "LunarCalendar",
+        "holidays",
+        "tqdm",
+        "cmdstanpy",
+        "matplotlib",
+        "setuptools-git",
+        "python-dateutil",
+        "pystan",
+        "prophet",
+    ]
```

Comment:
Can we just include `prophet`?

Reply:
Unfortunately, if we remove all of the dependencies for prophet it simply won't build. I went through exhaustive testing by removing each one, but the hard-dependencies in pystan's install on Cython mean that it won't build (gcc-g++ errors) if Cython isn't present before package installation and build. Similarly, if all of the other packages aren't already in the path prior to prophet install it won't install at all.

Reply:
+1! Perhaps an inline comment would help here to clarify for future readers.

Reply:
added comments in both locations with a note for a future pseudo TODO (if the build process gets refactored for Prophet, then remove all of the declared dependencies)